src/test/kotlin/uk/gov/justice/digital/hmpps/sendlegalmailtoprisonsapi/barcode/BarcodeServiceTest.kt

```diff
@@ -59,33 +63,42 @@ class BarcodeServiceTest {
     @Test
     fun `should throw not found and create barcode checked event if barcode does not exist`() {
       val expectedException = EntityNotFoundException("some_message")
-      whenever(barcodeRepository.findById("SOME_BARCODE")).thenReturn(Optional.empty())
-      whenever(barcodeRepository.save(any())).thenReturn(Barcode("SOME_BARCODE"))
-      whenever(barcodeEventService.checkForCreated(Barcode("SOME_BARCODE")))
-        .thenThrow(expectedException)
+      mockFindBarcode(null)
+      mockSaveBarcode()
+      whenever(barcodeEventService.checkForCreated(aBarcode())).thenThrow(expectedException)
 
       assertThatThrownBy { barcodeService.checkBarcode("some_user", "SOME_BARCODE", "some_location") }
         .isEqualTo(expectedException)
 
       verify(barcodeEventService)
-        .createEvent(Barcode("SOME_BARCODE"), "some_user", BarcodeStatus.CHECKED, "some_location")
+        .createEvent(aBarcode(), "some_user", BarcodeStatus.CHECKED, "some_location")
     }
   }
 
   @Nested
   inner class CheckBarcodeDuplicate {
     @Test
     fun `should throw validation exception and create checked event if duplicate`() {
-      val expectedException = ValidationException(Duplicate(Instant.now().minus(1, ChronoUnit.DAYS), "previous_location"))
-      whenever(barcodeRepository.findById("SOME_BARCODE")).thenReturn(Optional.of(Barcode("SOME_BARCODE")))
-      whenever(barcodeEventService.checkForDuplicate(Barcode("SOME_BARCODE"), "current_user", "current_location"))
+      val yesterday = Instant.now().minus(1, ChronoUnit.DAYS)
+      val expectedException = ValidationException(Duplicate(yesterday, "previous_location"))
+      mockFindBarcode()
+      whenever(barcodeEventService.checkForDuplicate(aBarcode(), "current_user", "current_location"))
         .thenThrow(expectedException)
 
       assertThatThrownBy { barcodeService.checkBarcode("current_user", "SOME_BARCODE", "current_location") }
         .isEqualTo(expectedException)
 
       verify(barcodeEventService)
-        .createEvent(Barcode("SOME_BARCODE"), "current_user", BarcodeStatus.CHECKED, "current_location")
+        .createEvent(aBarcode(), "current_user", BarcodeStatus.CHECKED, "current_location")
     }
   }
+
+  // Test helpers
+  private fun aBarcode() = Barcode("SOME_BARCODE")
```

Comment:
Given that this method is identical to the one in BarcodeEventServiceTest, would it be worth factoring this (and aBarcodeEvent from BarcodeEventServiceTest) into something that can be re-used by other tests?

In previous projects we had a concept of test data factories. They weren't 'factories' in the strictest sense of the design pattern, but basically were classes with public static methods that effectively did what you are doing here. (they were java projects rather that kotlin, hence static methods, but you get the idea)
Typically we'd have a package `testsupport` or similar, and we'd have subpackages for test data factories like this, and packages for other test support gubbins.

Reply:
As discussed I'm not keen on test factories due to the levels of indirection it introduces, but where there is a need I'd rather keep them in the same test class to minimise indirection. That said, if we find ourselves duplicating these helpers in more places then we should review and consider the benefits of a testsupport package.